### PR TITLE
feat: Inject metrics registry in route controllers

### PR DESCRIPTION
**Pull request by filvecchiato**
This PR allows for route-specific metrics to be emitted from the controllers, by injecting in the Response object (in locals) the metrics registry, making it available for each request received in case the prometheus flag is true

### CODE REVIEW

File: src/controllers/blocks/BlocksController.ts [hunk 1]
```diff
@@ -18,6 +18,7 @@ import type { ApiPromise } from '@polkadot/api';
 import { isHex } from '@polkadot/util';
 import { RequestHandler } from 'express';
 import { BadRequest } from 'http-errors';
+import client from 'prom-client';
```
Comment (TarikGul) [inline src/controllers/blocks/BlocksController.ts:21]: If we are using `client` here just as a type - lets do `import type client from ''prom-client`

File: src/controllers/blocks/BlocksController.ts [hunk 2]
```diff
@@ -124,13 +125,30 @@ export default class BlocksController extends AbstractController<BlocksService>
 	 * @param _req Express Request
 	 * @param res Express Response
 	 */
-	private getLatestBlock: RequestHandler = async (
-		{ query: { eventDocs, extrinsicDocs, finalized, noFees, decodedXcmMsgs, paraId } },
+	private getLatestBlock: RequestHandler<
```
Comment (TarikGul) [inline src/controllers/blocks/BlocksController.ts:128]: I think we can abstract this into `src/types/request.ts`

File: src/controllers/blocks/BlocksController.ts [hunk 3]
```diff
@@ -151,8 +169,7 @@ export default class BlocksController extends AbstractController<BlocksService>
 		}
 		const noFeesArg = noFees === 'true';
 		const decodedXcmMsgsArg = decodedXcmMsgs === 'true';
-		const paraIdArg =
-			paraId !== undefined ? this.parseNumberOrThrow(paraId as string, 'paraId must be an integer') : undefined;
+		const paraIdArg = paraId !== undefined ? this.parseNumberOrThrow(paraId, 'paraId must be an integer') : undefined;
```
Comment (TarikGul) [inline src/controllers/blocks/BlocksController.ts:172]:
```suggestion
		const paraIdArg = paraId  ? this.parseNumberOrThrow(paraId, 'paraId must be an integer') : undefined;
```

File: src/controllers/blocks/BlocksController.ts [hunk 4]
```diff
@@ -176,12 +221,34 @@ export default class BlocksController extends AbstractController<BlocksService>
 	 * @param req Express Request
 	 * @param res Express Response
 	 */
-	private getBlockById: RequestHandler<INumberParam> = async (
-		{ params: { number }, query: { eventDocs, extrinsicDocs, noFees, finalizedKey, decodedXcmMsgs, paraId } },
+	private getBlockById: RequestHandler<
```
[the remark this answers is not in the document]
Comment (TarikGul) [reply src/controllers/blocks/BlocksController.ts:224]: We can also make this Generic so you don't have to write it twice:

type IBlockRequestHandler<T> = {
  T,
  unknown,
  unknown,
  BlockQueryParams,
  Metrics
}

File: src/controllers/blocks/BlocksController.ts [hunk 5]
```diff
@@ -250,10 +341,18 @@ export default class BlocksController extends AbstractController<BlocksService>
 	 * @param req Express Request
 	 * @param res Express Response
 	 */
-	private getBlocks: RequestHandler<unknown, unknown, unknown, IRangeQueryParam> = async (
-		{ query: { range, eventDocs, extrinsicDocs, noFees } },
-		res,
-	): Promise<void> => {
+	private getBlocks: RequestHandler<
```
Comment (TarikGul) [inline src/controllers/blocks/BlocksController.ts:344]: Same with this type

File: src/controllers/blocks/BlocksRawExtrinsicsController.ts [hunk 6]
```diff
@@ -41,9 +42,32 @@ export default class BlocksRawExtrinsicsController extends AbstractController<Bl
 	 * @param _req Express Request
 	 * @param res Express Response
 	 */
-	private getBlockRawExtrinsics: RequestHandler<INumberParam> = async ({ params: { blockId } }, res): Promise<void> => {
+	private getBlockRawExtrinsics: RequestHandler<
```
Comment (TarikGul) [inline src/controllers/blocks/BlocksRawExtrinsicsController.ts:45]: Same with this Type

**Review by TarikGul — CHANGES_REQUESTED** [review]
Overall, really great job. Just some nits for the types.

File: src/controllers/blocks/BlocksController.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (filvecchiato) [reply src/controllers/blocks/BlocksController.ts]: Created a generic type for the request with metrics that extends the RequestHandler

**Review by TarikGul — APPROVED** [review]
Great job, LGTM 🚀

**Review by Imod7 — APPROVED** [review]
LGTM! Just two small comments.

File: src/controllers/blocks/BlocksController.ts [hunk 7]
```diff
@@ -167,7 +166,35 @@ export default class BlocksController extends AbstractController<BlocksService>
...
+				.labels({ method: method, route: path, status_code: res.statusCode })
+				.observe(block.extrinsics.length);
+
+			const extrinsics_per_second = res.locals.metrics.registry['sas_extrinsics_per_second_count'] as client.Histogram;
+			extrinsics_per_second
+				.labels({ method: method, route: path, status_code: res.statusCode })
+				.observe(block.extrinsics.length / seconds);
+
+			const extrinsicsPerBlockMetrics = res.locals.metrics.registry[
+				'sas_extrinsics_per_block_count'
```
[the remark this answers is not in the document]
Comment (filvecchiato) [reply src/controllers/blocks/BlocksController.ts]: Nice catch, I renamed them!

**Pull request merged**